internal/transport/http_util.go

```diff
@@ -389,7 +408,20 @@ func newFramer(conn net.Conn, writeBufferSize, readBufferSize int, maxHeaderList
 	if readBufferSize > 0 {
 		r = bufio.NewReaderSize(r, readBufferSize)
 	}
-	w := newBufWriter(conn, writeBufferSize)
+	writeBufferMutex.Lock()
+	size := writeBufferSize * 2
+	pool, ok := writeBufferPoolMap[size]
+	if !ok {
+		pool = &sync.Pool{
+			New: func() interface{} {
+				b := make([]byte, size)
+				return &b
+			},
+		}
+		writeBufferPoolMap[size] = pool
+	}
+	writeBufferMutex.Unlock()
```

Comment:
In theory, we can get rid of this mutext and pool map if we store sync.Pool per ClientConnection and per server. However, this will require modifying a lot of internal interface to pass the pool all the way down to the transport layer. This also will prevent us from sharing the pool between different ClientConnections and servers, so I decided to take the mutex approach. 

As far as I understand, this mutex shouldn't really impact the performance, as it got called only on new connection creation, which happens not very often in http2.

Reply:
When do entries from this map ever get deleted? Can this map grow in a unbounded fashion?

Reply:
Right now - never. In theory it can grow unbounded if the app creates multiple servers and/or ClientConnections with different buffer sizes. I can't think about any use-case where this could be useful, but it is possible. I am not sure if it worth adding more complicated logic to protect against this, because my understanding is that go runtime will eventually deallocate all memory from unused pools, so we will be waisting memory only on storing pointers to empty pools, which should be neglectable in practise. 

Still I can fix this if you prefer:
* I can add a goroutine that removes unused pools. I'll probably need to store something like lastModifedTime for each pool to do that.  The downsides are some additional complexity + the fact that we need to take the lock in the goroutine.
* I can store a pool per Server and ClientConnection and propagate it all the way down to the transport layer.  The downsides are the fact that we need to extend some internal interfaces + we won't be able to reuse pools between different connections that uses the same buffer size (which I assume is a very common use-case) But in this case we can get rid of the lock completely.

Do you have any other ideas? Let me know what I should do here.

Reply:
> In theory it can grow unbounded if the app creates multiple servers and/or ClientConnections with different buffer sizes. I can't think about any use-case where this could be useful, but it is possible.

I agree. I don't expect applications to create multiple `grpc.ClientConn`s or `grpc.Server`s and pass them different values for the send and receive buffer sizes. But given that we are going to have a dial option and a server option to enable the sync.Pool on the client and server respectively, we could document it there, just so no one complains saying a few bytes are leaked.

@dfawley : Thoughts/Objections?